### PR TITLE
Simplified CxxStdlib

### PR DESCRIPTION
**Pull request by koraktor**
This drops the polymorphism introduced in e1f97e260ca320141c9c48158814aff895e76486 which isn’t really used here.

/cc @jacknagel

### CODE REVIEW

**Comment by jacknagel** [comment]
I don't work on Homebrew anymore, but the rationale for using polymorphism here is that we only have to perform the regex match once, at instantiation time, rather than on each comparison. It's purely a design choice. Why change it?

File: Library/Homebrew/cxxstdlib.rb [hunk 1]
```diff
@@ -16,20 +16,16 @@ def self.create(type, compiler)
...
     stdlib = create(keg.detect_cxx_stdlibs.first, compiler)
-
-    begin
-      stdlib.check_dependencies(formula, deps)
-    rescue CompatibilityError => e
-      opoo e.message
-    end
+    stdlib.check_dependencies(formula, deps)
+  rescue CompatibilityError
+    opoo $!.message
```
Comment (jacknagel) [inline Library/Homebrew/cxxstdlib.rb]: Why change the error handling? The new version is less correct (the scope of the rescue has changed). What is the point of switching from `=> e` to `$?`?).

**Comment by koraktor** [comment]
IMHO it’s harder to understand code that uses two different subclasses just to represent `true` / `false`. If the performance is really an issue here, there could be a boolean `@apple_compiler` variable instead.

**Comment by DomT4** [comment]
Going to pass on this for now. If Jack isn't enthusiastic about the changed syntax around compiler handling you'll struggle to find another maintainer who is. Unless there's a significant performance issue here not sure what the absolute upside would be, and we've been slowly getting API documentation sorted to make the code clearer. Thanks for the submission though! :heart:

**Pull request closed without merging**